src/core.jl

```diff
@@ -132,6 +132,7 @@ function serve(ctx::Context;
         if ctx.mod === nothing
             @warn "You are trying to use the `revise` option without @oxidise. Code in the `Main` module, which likely includes your routes, will not be tracked and revised."
         end
+        middleware = convert(Vector{Any}, middleware)
```

Comment:
rather than performing a type conversion on the entire array, could we try setting the middleware keyword type to  `Vector{Function}`. This should be generic enough to handle any weird edge cases

Reply:
I think doing that will mean that middleware can't be implemented as functors/callable structs. Since Julia doesn't have interfaces/traits there's no way to get the type `Union[Callable Structs, Function}`.

e.g.

```
function testy(f::Vector{Function})
    for i in f
        i(1)
    end
end

struct MyFunctor end

function (f::MyFunctor)(x)
    return x + 1
end

testy([MyFunctor()])
```

`ERROR: MethodError: no method matching testy(::Vector{MyFunctor})`

Dispatching on a more general parameter won't work since Julia containers as type invariant:

```
function testy2(f::Vector{Any})
    for i in f
        i(1)
    end
end

testy2([MyFunctor()])
```

`ERROR: MethodError: no method matching testy2(::Vector{MyFunctor})`

The latter could be made to work, but it means that users have to pass middleware as: `middleware=Any[mymiddleware1]` if there is a single middleware.

Reply:
@frankier, 

I think that's actually more of a feature than a limiting factor. The underlying implementation by HTTP only accepts higher ordered functions as valid middleware. While technically functors could be used for middleware - I'm not sure I'd want to enable that. 

What scenarios do you see this being more helpful than a regular HOF?

Reply:
When making a functor (C++ term) or callable (Python term) the idea is to override the "function-pointer call" operator, so that the struct behaves like a function pointer. In practice it can be useful e.g. when you want to have multiple initialisation paths. Another possibility is that you may want to add extra methods to the middleware if it exists as part of a larger flow.

It is true that things like this are not strictly necessary. One can always wrap ones callable in a HOF, but then why should one have to? Accepting both functions and HOF is part of conventional Julia interfaces.

It currently works totally fine with both HTTP.jl and Oxygen.jl, e.g. I have been using the following middleware for profiling:

```
using Profile
using ProfileSVG
using Oxygen
using HTTP


@kwdef mutable struct FlamegraphMiddleware
    profidx::Int = 1
    lock::ReentrantLock = ReentrantLock()
    profilesvg_kwargs=(;)
end


function (middleware::FlamegraphMiddleware)(handler)
    return function(req::HTTP.Request)
        Profile.clear()
        try
            return @profile handler(req)
        finally
            mkpath("profiles")
            data = Profile.fetch()
            if isempty(data)
                @info "No profile data (try increasing the sample rate)"
            else
                Profile.clear()
                local dest
                lock(middleware.lock)
                try
                    dest = joinpath("profiles", "$(middleware.profidx).svg")
                    middleware.profidx += 1
                finally
                    unlock(middleware.lock)
                end
                rm(dest; force=true)
                ProfileSVG.save(dest, data; middleware.profilesvg_kwargs...)
                @info "Profile saved to $dest with $(length(data)) samples"
            end
        end
    end
end
```

I think its nice that both HOF and struct middleware work fine.

Types in function signatures are usually mainly for dispatch in Julia. If you pass something of the wrong type, you'll just get a MethodError, and Julia specializes at call-time. I think functions are traditionally an exception to this and you have to specifically add in an unbounded type variable for specialisation. See https://docs.julialang.org/en/v1/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing

Reply:
It does seem that HTTP.jl is duck-typed too https://github.com/JuliaWeb/HTTP.jl/blob/master/src/Handlers.jl#L346

Reply:
Ok, in that case it makes sense to keep it open if HTTP allows it